### PR TITLE
Add optimization for multiple filters in `where` for documents API

### PR DESCRIPTION
**Pull request by EricBorczuk**
This changes the format of the paging state for the documents API, implementing a "last seen" document ID that is used during search requests.

This allows us to optimize searches with multiple AND's a bit, whose implementation is in this PR.

### CODE REVIEW

**Review by dimas-b — APPROVED** [review]

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java [hunk 1]
```diff
@@ -999,8 +993,6 @@ void addRowsToMap(Map<String, List<Row>> rowsByDoc, List<Row> rows) {
     List<BuiltCondition> predicate =
         ImmutableList.of(BuiltCondition.of("key", Predicate.IN, new ArrayList<>(docNames)));
 
-    db = dbFactory.getDocDataStoreForToken(authToken, headers);
```
Comment (mpenick) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:1002]: Was this leftover from something else? Or is there some other reason?

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java [hunk 2]
```diff
@@ -891,41 +888,38 @@ void addRowsToMap(Map<String, List<Row>> rowsByDoc, List<Row> rows) {
...
-      // CQL (multiple C* filters would indirectly require an `OR` due to the document schema
-      // structure).
-      inCassandraFilters = Collections.emptyList();
-      inMemoryFilters = filters;
+            .findFirst();
+    List<FilterCondition> inCassandraFilters = new ArrayList<>();
+    List<FilterCondition> inMemoryFilters = filters;
+
+    String cassandraPath;
+    if (firstSupportedFilter.isPresent()) {
```
Comment (mpenick) [inline restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:912]: So is the optimization that it uses the first supported filter on the server-side then filters the rest in memory? versus before where it just filtered everything in memory.

Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:912]: Exactly, yes!

File: restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java (hunk 1 above)
Comment (EricBorczuk) [reply restapi/src/main/java/io/stargate/web/docsapi/service/DocumentService.java:1002]: Yea this was just left over and didn't need to be here

**Comment by EricBorczuk** [comment]
Putting this on hold for now, if Calcite is going to replace this then this change becomes throwaway code

File: restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java [hunk 3]
```diff
@@ -13,25 +13,63 @@
...
+  @JsonIgnore boolean doneSkipping;
+
   @JsonCreator
   public DocumentSearchPageState(
       @JsonProperty("documentId") final String documentId,
       @JsonProperty("internalPageState") final String internalPageState) {
     this.documentId = documentId;
     this.internalPageState = internalPageState;
+    this.idFound = false;
+    this.doneSkipping = false;
```
Comment (dimas-b) [inline restapi/src/main/java/io/stargate/web/docsapi/dao/DocumentSearchPageState.java:29]: nit: `false` is the default for `boolean` fields.

**Review by dimas-b — APPROVED** [review]

**Comment by EricBorczuk** [comment]
In all, this change improves latency by a ton!

on master:
[READ], AverageLatency(us), 12312642
[READ], MinLatency(us), 8896512
[READ], MaxLatency(us), 15982591
[READ], 95thPercentileLatency(us), 14909439
[READ], 99thPercentileLatency(us), 15851519

with these changes:
[READ], AverageLatency(us), 15958.11994
[READ], MinLatency(us), 4940
[READ], MaxLatency(us), 970751
[READ], 95thPercentileLatency(us), 27023
[READ], 99thPercentileLatency(us), 40831

which is 771x faster

**Pull request merged**